src/palindrom.js

```diff
@@ -34,951 +33,1016 @@ const NodeWebSocket = require('websocket').w3cwebsocket;
+                    'Previous session ID: ' + this.remoteUrl,
+                    'New session ID: ' + remoteUrl
+                ].join('\n');
+
+                throw new PalindromError(message);
+            }
+            this.remoteUrlSet = true;
+            this.remoteUrl = new URL(remoteUrl, this.remoteUrl.href);
+        }
+
+        handleResponseHeader(res) {
+            /* Axios always returns lowercase headers */
+            const location =
+                res.headers &&
+                (res.headers['x-location'] || res.headers['location']);
+            if (location) {
+                this.setRemoteUrl(location);
+            }
+        }
+
+        /**
+         * Internal method to perform XMLHttpRequest
+         * @param url (Optional) URL to send the request. If empty string, undefined or null given - the request will be sent to window location
+         * @param accept (Optional) HTTP accept header
+         * @param data (Optional) Data payload
+         * @param [callback(response)] callback to be called in context of palindrom with response as argument
+         * @returns {XMLHttpRequest} performed XHR
+         */
+        xhr(url, accept, data, callback, setReferer) {
+            const method = data ? 'PATCH' : 'GET';
+            const headers = {};
+            let requestPromise;
+
+            if (data) {
+                headers['Content-Type'] = 'application/json-patch+json';
+            }
+            if (accept) {
+                headers['Accept'] = accept;
+            }
+            if (this.remoteUrl && setReferer) {
+                headers['X-Referer'] = this.remoteUrl.pathname;
+            }
+            if (method === 'GET') {
+                requestPromise = axios.get(url, {
+                    headers
+                });
+            } else {
+                requestPromise = axios.patch(url, data, {
+                    headers
+                });
+            }
+            requestPromise
+                .then(res => {
+                    this.handleResponseHeader(res);
+                    callback && callback.call(this.palindrom, res, method);
+                })
+                .catch(error => {
+                    const res = error.response;
+
+                    if (res) {
+                        var statusCode = res.status;
+                        var statusText = res.statusText || res.data;
+                        var reason = res.data;
+
+                        //this is not a fatal error
+                        if (
+                            statusCode >= 400 &&
+                            statusCode < 500 &&
+                            res.headers['content-type'] ===
+                                'application/json-patch+json'
+                        ) {
+                            this.handleResponseHeader(res);
+                            callback &&
+                                callback.call(this.palindrom, res, method);
```

Comment:
Sorry, I missed this because GitHub collapsed the file for me. Could you please make likes 539-541 and 521-522 dry? They are doing the same.